File: backend/src/deaths_door/routes/game.py
"""Routes for game management, night phases, and consolidated state."""

from uuid import UUID

from fastapi import APIRouter
from fastapi.exceptions import HTTPException
from pydantic import BaseModel, Field

from ..character import CharacterOut
from ..events import EventType, FirstNightSet, NightStepSet, describe_event
from ..game_manager import game_manager
from ..game_state import game_state_to_included_role_outs, player_state_to_out
from ..night_step import NightStep
from ..player import PlayerOut
from ..script_name import ScriptName
from ..status_effects import StatusEffectOut
from . import timer as timer_routes

router = APIRouter(prefix="/game", tags=["Game Management"])


class NewGameRequest(BaseModel):
    """Request to start a new game with a specific script."""

    script_name: str = Field(
        ...,
        description="Name of the script/edition to play",
        examples=["trouble_brewing", "bad_moon_rising", "sects_and_violets"],
    )


class NewGameResponse(BaseModel):
    """Response after creating a new game."""

    status: str = Field(..., description="Operation status", examples=["success"])
    script_name: str = Field(
        ...,
        description="Name of the script that was loaded",
        examples=["trouble_brewing"],
    )


@router.post("/new", responses={404: {"description": "Script not found"}})
async def new_game(req: NewGameRequest) -> NewGameResponse:
    """Start a new game with the specified script."""
    script_name = ScriptName.from_str(req.script_name)
    if script_name is None:
        raise HTTPException(status_code=404, detail="Script not found")

    await game_manager.create_game(script_name.value)
    return NewGameResponse(status="success", script_name=script_name.value)


@router.get("/script/name")
async def get_game_script() -> str:
    """Return the name of the script for the current game."""
    state = await game_manager.get_state()
    return state.script_name


@router.get("/script/roles")
async def get_game_script_roles() -> list[CharacterOut]:
    """
    Return all possible character roles available in the current script.

    This returns the complete set of roles defined by the script (e.g., all Trouble Brewing roles).
    To see which roles have been added to the current game, use GET /characters/list instead.
    """
    state = await game_manager.get_state()
    script = state.get_script()
    return [c.to_out() for c in script.characters]


@router.get("/script/night/first")
async def get_game_first_night_steps() -> list[NightStep]:
    """Return the first night steps."""
    state = await game_manager.get_state()
    # Temporarily set first night to get those steps
    first_night_state = state.model_copy(update={"is_first_night": True})
    return first_night_state.get_night_steps()


@router.get("/script/night/other")
async def get_game_other_night_steps() -> list[NightStep]:
    """Return the other night steps (subsequent nights)."""
    state = await game_manager.get_state()
    other_night_state = state.model_copy(update={"is_first_night": False})
    return other_night_state.get_night_steps()


@router.get("/status_effects")
async def get_game_status_effects() -> list[StatusEffectOut]:
    """Return the status effects for the current game."""
    state = await game_manager.get_state()
    return state.get_status_effects()


class GameStateResponse(BaseModel):
    """Complete game state including all relevant information."""

    script_name: str = Field(
        ...,
        description="Name of the current script/edition",
        examples=["trouble_brewing"],
    )
    players: list[PlayerOut] = Field(
        ...,
        description="All players in the game with their characters and status",
    )
    current_night_step: str = Field(
        ...,
        description="Current night step bookmark",
        examples=["Dusk", "Poisoner", "Imp"],
    )
    is_first_night: bool = Field(
        ...,
        description="Whether this is the first night or a subsequent night",
    )
    should_reveal_roles: bool = Field(
        ...,
        description="Whether roles should be revealed to players",
    )
    status_effects: list[StatusEffectOut] = Field(
        ...,
        description="All active status effects in the game",
    )
    included_roles: list[CharacterOut] = Field(
        ...,
        description="Roles that have been included but not yet assigned to players",
    )
    night_steps: list[NightStep] = Field(
        ...,
        description="Night steps for the current night (filtered based on is_first_night)",
    )
    living_player_count: int = Field(
        ...,
        description="Number of living players",
    )
    execution_threshold: int = Field(
        ...,
        description="Number of votes needed to execute (≥50% of living players)",
    )
    dead_players_with_vote: list[str] = Field(
        ...,
        description="Names of dead players who haven't used their dead vote",
    )
    timer: timer_routes.TimerStateResponse = Field(
        ...,
        description="Current timer state (is_running and seconds remaining)",
    )


@router.get("/state")
async def get_game_state() -> GameStateResponse:
    """Get the complete game state in a single request."""
    state = await game_manager.get_state()
    script = state.get_script()

    # Get timer state
    timer_is_running = await timer_routes.state.get_is_running()
    timer_seconds = await timer_routes.state.get_seconds()

    return GameStateResponse(
        script_name=state.script_name,
        players=[player_state_to_out(p, script) for p in state.players],
        current_night_step=state.current_night_step,
        is_first_night=state.is_first_night,
        should_reveal_roles=state.should_reveal_roles,
        status_effects=state.get_status_effects(),
        included_roles=game_state_to_included_role_outs(state),
        night_steps=state.get_night_steps(),
        living_player_count=state.living_player_count,
        execution_threshold=state.execution_threshold,
        dead_players_with_vote=state.get_dead_players_with_vote(),
        timer=timer_routes.TimerStateResponse(
            is_running=timer_is_running,
            seconds=timer_seconds,
        ),
    )


class NightPhaseResponse(BaseModel):
    """Response containing current night phase information."""

    current_night_step: str = Field(
        ...,
        description="Name of the current night step (e.g., 'Dusk', 'Poisoner')",
        examples=["Dusk", "Poisoner", "Fortune Teller"],
    )
    is_first_night: bool = Field(
        ...,
        description="Whether this is the first night or a subsequent night",
        examples=[True, False],
    )


@router.get("/night/phase")
async def get_night_phase() -> NightPhaseResponse:
    """Get the current night phase information."""
    state = await game_manager.get_state()
    return NightPhaseResponse(
        current_night_step=state.current_night_step,
        is_first_night=state.is_first_night,
    )


class SetNightStepRequest(BaseModel):
    """Request to set the current night step."""

    step: str = Field(
        ...,
        description="Name of the night step to set as current",
        examples=["Dusk", "Poisoner", "Imp", "Dawn"],
    )


@router.post("/night/phase/step")
async def set_night_step(req: SetNightStepRequest) -> NightPhaseResponse:
    """Set the current night step."""
    new_state = await game_manager.dispatch(NightStepSet(step=req.step))
    return NightPhaseResponse(
        current_night_step=new_state.current_night_step,
        is_first_night=new_state.is_first_night,
    )


class SetFirstNightRequest(BaseModel):
    """Request to set whether it's the first night."""

    is_first_night: bool = Field(
        ...,
        description="Whether this is the first night",
        examples=[True, False],
    )


@router.post("/night/phase/first_night")
async def set_first_night(req: SetFirstNightRequest) -> NightPhaseResponse:
    """Set whether this is the first night (automatically resets to Dusk)."""
    new_state = await game_manager.dispatch(FirstNightSet(is_first_night=req.is_first_night))
    return NightPhaseResponse(
        current_night_step=new_state.current_night_step,
        is_first_night=new_state.is_first_night,
    )


@router.get("/script/night/steps")
async def get_night_steps() -> list[NightStep]:
    """Return the night steps for the current night (first or other based on game state)."""
    state = await game_manager.get_state()
    return state.get_night_steps()


# --- Event sourcing endpoints ---


class EventOut(BaseModel):
    """A game event for the history endpoint."""

    version: int = Field(..., description="Version after this event (1-indexed)")
    description: str = Field(..., description="Human-readable description of the event")
    event_type: EventType = Field(..., description="Type of event")
    timestamp: str = Field(..., description="ISO 8601 timestamp")
    payload: dict[str, object] = Field(..., description="Event payload data")


class HistoryResponse(BaseModel):
    """Response containing the event history for the current game."""

    game_id: str = Field(..., description="Current game ID")
    version: int = Field(..., description="Current version (number of events applied)")
    events: list[EventOut] = Field(..., description="All events in chronological order")


@router.get("/history")
async def get_game_history() -> HistoryResponse:
    """Get the full event history for the current game."""
    state = await game_manager.get_state()
    events = await game_manager.get_history()
    return HistoryResponse(
        game_id=str(state.game_id),
        version=state.version,
        events=[
            EventOut(
                version=e.sequence + 1,
                description=describe_event(e.payload),
                event_type=e.payload.type,
                timestamp=e.timestamp.isoformat(),
                payload=e.payload.model_dump(exclude={"type"}),
            )
            for e in events
        ],
    )


class RewindRequest(BaseModel):
    """Request to rewind the game to a specific version."""

    to_version: int = Field(..., description="Version to rewind to (1-based, inclusive)", ge=1)


@router.post("/rewind", responses={400: {"description": "Invalid version"}})
async def rewind_game(req: RewindRequest) -> GameStateResponse:
    """Rewind the current game to a previous version, deleting subsequent events."""
    try:
        await game_manager.rewind(req.to_version)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e

    return await get_game_state()


class ForkRequest(BaseModel):
    """Request to fork the game from a specific version."""

    from_version: int = Field(..., description="Version to fork from (1-based, inclusive)", ge=1)


class ForkResponse(BaseModel):
    """Response after forking a game."""

    new_game_id: str = Field(..., description="ID of the newly forked game")
    version: int = Field(..., description="Version of the forked game")


@router.post("/fork", responses={400: {"description": "Invalid version"}})
async def fork_game(req: ForkRequest) -> ForkResponse:
    """Fork the current game from a specific version, creating a new game branch."""
    try:
        new_state = await game_manager.fork(req.from_version)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e

    return ForkResponse(new_game_id=str(new_state.game_id), version=new_state.version)


class LoadGameRequest(BaseModel):
    """Request to load a game by ID."""

    game_id: str = Field(..., description="UUID of the game to load")


@router.post("/load", responses={400: {"description": "Game not found"}})
async def load_game(req: LoadGameRequest) -> GameStateResponse:
    """Load a previously saved game by its ID."""
    try:
        game_id = UUID(req.game_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=f"Invalid game ID: {req.game_id}") from e

    try:
        await game_manager.load_game(game_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e

    return await get_game_state()


class GameListResponse(BaseModel):
    """Response listing all saved games."""

    game_ids: list[str] = Field(..., description="List of all game IDs")


@router.get("/list")
async def list_games() -> GameListResponse:
    """List all saved game IDs."""
    ids = await game_manager.list_games()
    return GameListResponse(game_ids=[str(gid) for gid in ids])

File: backend-rust/src/routes/game.rs
//! Routes for game management, night phases, and consolidated state.

use std::convert::Infallible;

use axum::extract::State;
use axum::response::sse::{Event, KeepAlive, Sse};
use axum::Json;
use futures_core::Stream;
use serde::{Deserialize, Serialize};
use tokio_stream::wrappers::BroadcastStream;
use tokio_stream::StreamExt;
use utoipa::ToSchema;
use utoipa_axum::router::OpenApiRouter;
use utoipa_axum::routes;
use uuid::Uuid;

use crate::app::{AppError, AppJson, AppResult, AppState};
use crate::character::CharacterOut;
use crate::events::{describe_event, EventPayload};
use crate::game_state::{game_state_to_included_role_outs, player_state_to_out, GameState};
use crate::night_step::NightStep;
use crate::player::PlayerOut;
use crate::routes::timer::TimerStateResponse;
use crate::script_name::ScriptName;
use crate::status_effect::StatusEffectOut;

pub fn router() -> OpenApiRouter<AppState> {
    OpenApiRouter::new()
        .routes(routes!(new_game))
        .routes(routes!(get_game_script))
        .routes(routes!(get_game_script_roles))
        .routes(routes!(get_first_night_steps))
        .routes(routes!(get_other_night_steps))
        .routes(routes!(get_night_steps))
        .routes(routes!(get_status_effects))
        .routes(routes!(get_game_state))
        .routes(routes!(get_night_phase))
        .routes(routes!(set_night_step))
        .routes(routes!(set_first_night))
        .routes(routes!(get_game_history))
        .routes(routes!(rewind_game))
        .routes(routes!(fork_game))
        .routes(routes!(load_game))
        .routes(routes!(list_games))
        .routes(routes!(stream_game_state))
}

// --- Shared state serialization ---

/// The full game-state snapshot pushed over SSE and embedded in `/game/state`.
#[derive(Serialize, ToSchema)]
pub struct GameStateBase {
    pub script_name: String,
    pub players: Vec<PlayerOut>,
    /// Current night-step bookmark, e.g. "Dusk" or "Imp".
    pub current_night_step: String,
    pub is_first_night: bool,
    pub should_reveal_roles: bool,
    pub status_effects: Vec<StatusEffectOut>,
    /// Roles in the pool not yet assigned to a player.
    pub included_roles: Vec<CharacterOut>,
    /// Night steps for the current phase, filtered to roles in play.
    pub night_steps: Vec<NightStep>,
    pub living_player_count: usize,
    /// Votes needed to execute (>= 50% of living players).
    pub execution_threshold: usize,
    pub dead_players_with_vote: Vec<String>,
}

pub fn game_state_base(state: &GameState) -> GameStateBase {
    let players = match state.get_script() {
        Some(script) => state
            .players
            .iter()
            .map(|p| player_state_to_out(p, script))
            .collect(),
        None => Vec::new(),
    };
    GameStateBase {
        script_name: state.script_name.clone(),
        players,
        current_night_step: state.current_night_step.clone(),
        is_first_night: state.is_first_night,
        should_reveal_roles: state.should_reveal_roles,
        status_effects: state.get_status_effects(),
        included_roles: game_state_to_included_role_outs(state),
        night_steps: state.get_night_steps(),
        living_player_count: state.living_player_count(),
        execution_threshold: state.execution_threshold(),
        dead_players_with_vote: state.get_dead_players_with_vote(),
    }
}

/// Full `/game/state` response: the snapshot plus the timer.
#[derive(Serialize, ToSchema)]
pub struct GameStateResponse {
    #[serde(flatten)]
    pub base: GameStateBase,
    pub timer: TimerStateResponse,
}

pub async fn build_game_state_response(state: &AppState) -> AppResult<GameStateResponse> {
    let game = state.manager.get_state().await?;
    let timer = TimerStateResponse {
        is_running: state.timer.get_is_running().await,
        seconds: state.timer.get_seconds().await,
    };
    Ok(GameStateResponse {
        base: game_state_base(&game),
        timer,
    })
}

fn sse_payload(state: &GameState) -> String {
    serde_json::to_string(&game_state_base(state)).unwrap_or_else(|_| "{}".to_string())
}

// --- New game ---

#[derive(Deserialize, ToSchema)]
pub struct NewGameRequest {
    /// Script/edition id, e.g. `trouble_brewing`.
    #[schema(example = "trouble_brewing")]
    pub script_name: String,
}

#[derive(Serialize, ToSchema)]
pub struct NewGameResponse {
    pub status: String,
    pub script_name: String,
}

/// Start a new game with the given script.
///
/// The new game starts with **no roles** in the pool — add them via
/// `POST /characters/add/multi` before adding players.
#[utoipa::path(
    post, path = "/game/new", tag = "Game",
    request_body = NewGameRequest,
    responses(
        (status = 200, description = "Game created", body = NewGameResponse),
        (status = 404, description = "Unknown script")
    )
)]
async fn new_game(
    State(state): State<AppState>,
    AppJson(req): AppJson<NewGameRequest>,
) -> AppResult<Json<NewGameResponse>> {
    let script_name = ScriptName::from_str(&req.script_name)
        .ok_or_else(|| AppError::not_found("Script not found"))?;
    state.manager.create_game(script_name.value()).await?;
    Ok(Json(NewGameResponse {
        status: "success".to_string(),
        script_name: script_name.value().to_string(),
    }))
}

/// The current game's script id.
#[utoipa::path(get, path = "/game/script/name", tag = "Game", responses((status = 200, body = String)))]
async fn get_game_script(State(state): State<AppState>) -> AppResult<Json<String>> {
    let game = state.manager.get_state().await?;
    Ok(Json(game.script_name))
}

/// All character roles defined by the current script (the full catalog, not the
/// game's pool — for that, use `GET /characters/list`).
#[utoipa::path(get, path = "/game/script/roles", tag = "Game", responses((status = 200, body = [CharacterOut])))]
async fn get_game_script_roles(
    State(state): State<AppState>,
) -> AppResult<Json<Vec<CharacterOut>>> {
    let game = state.manager.get_state().await?;
    let roles = match game.get_script() {
        Some(script) => script.characters.iter().map(|c| c.to_out()).collect(),
        None => Vec::new(),
    };
    Ok(Json(roles))
}

/// First-night steps for the current game (filtered to roles in play).
#[utoipa::path(get, path = "/game/script/night/first", tag = "Game", responses((status = 200, body = [NightStep])))]
async fn get_first_night_steps(State(state): State<AppState>) -> AppResult<Json<Vec<NightStep>>> {
    let mut game = state.manager.get_state().await?;
    game.is_first_night = true;
    Ok(Json(game.get_night_steps()))
}

/// Subsequent-night steps for the current game (filtered to roles in play).
#[utoipa::path(get, path = "/game/script/night/other", tag = "Game", responses((status = 200, body = [NightStep])))]
async fn get_other_night_steps(State(state): State<AppState>) -> AppResult<Json<Vec<NightStep>>> {
    let mut game = state.manager.get_state().await?;
    game.is_first_night = false;
    Ok(Json(game.get_night_steps()))
}

/// Night steps for the current phase (first vs. other, based on game state).
#[utoipa::path(get, path = "/game/script/night/steps", tag = "Game", responses((status = 200, body = [NightStep])))]
async fn get_night_steps(State(state): State<AppState>) -> AppResult<Json<Vec<NightStep>>> {
    let game = state.manager.get_state().await?;
    Ok(Json(game.get_night_steps()))
}

/// All status effects currently introduced by characters in the game.
#[utoipa::path(get, path = "/game/status_effects", tag = "Game", responses((status = 200, body = [StatusEffectOut])))]
async fn get_status_effects(
    State(state): State<AppState>,
) -> AppResult<Json<Vec<StatusEffectOut>>> {
    let game = state.manager.get_state().await?;
    Ok(Json(game.get_status_effects()))
}

/// The complete game state in one request (players, night phase, vote info,
/// status effects, and timer). For live updates, prefer the SSE `/game/stream`.
#[utoipa::path(get, path = "/game/state", tag = "Game", responses((status = 200, body = GameStateResponse)))]
async fn get_game_state(State(state): State<AppState>) -> AppResult<Json<GameStateResponse>> {
    Ok(Json(build_game_state_response(&state).await?))
}

// --- Night phase ---

#[derive(Serialize, ToSchema)]
pub struct NightPhaseResponse {
    pub current_night_step: String,
    pub is_first_night: bool,
}

/// The current night phase (step bookmark + first-night flag).
#[utoipa::path(get, path = "/game/night/phase", tag = "Game", responses((status = 200, body = NightPhaseResponse)))]
async fn get_night_phase(State(state): State<AppState>) -> AppResult<Json<NightPhaseResponse>> {
    let game = state.manager.get_state().await?;
    Ok(Json(NightPhaseResponse {
        current_night_step: game.current_night_step,
        is_first_night: game.is_first_night,
    }))
}

#[derive(Deserialize, ToSchema)]
pub struct SetNightStepRequest {
    /// Night-step name to bookmark, e.g. "Poisoner".
    #[schema(example = "Poisoner")]
    pub step: String,
}

/// Set the current night-step bookmark.
#[utoipa::path(
    post, path = "/game/night/phase/step", tag = "Game",
    request_body = SetNightStepRequest,
    responses((status = 200, body = NightPhaseResponse))
)]
async fn set_night_step(
    State(state): State<AppState>,
    AppJson(req): AppJson<SetNightStepRequest>,
) -> AppResult<Json<NightPhaseResponse>> {
    let new_state = state
        .manager
        .dispatch(EventPayload::NightStepSet { step: req.step })
        .await?;
    Ok(Json(NightPhaseResponse {
        current_night_step: new_state.current_night_step,
        is_first_night: new_state.is_first_night,
    }))
}

#[derive(Deserialize, ToSchema)]
pub struct SetFirstNightRequest {
    pub is_first_night: bool,
}

/// Toggle the first-night flag (resets the step bookmark to "Dusk").
#[utoipa::path(
    post, path = "/game/night/phase/first_night", tag = "Game",
    request_body = SetFirstNightRequest,
    responses((status = 200, body = NightPhaseResponse))
)]
async fn set_first_night(
    State(state): State<AppState>,
    AppJson(req): AppJson<SetFirstNightRequest>,
) -> AppResult<Json<NightPhaseResponse>> {
    let new_state = state
        .manager
        .dispatch(EventPayload::FirstNightSet {
            is_first_night: req.is_first_night,
        })
        .await?;
    Ok(Json(NightPhaseResponse {
        current_night_step: new_state.current_night_step,
        is_first_night: new_state.is_first_night,
    }))
}

// --- Event sourcing endpoints ---

#[derive(Serialize, ToSchema)]
pub struct EventOut {
    /// 1-indexed version produced by this event (use with rewind/fork).
    pub version: i64,
    pub description: String,
    pub event_type: String,
    pub timestamp: String,
    /// Event-specific payload (fields vary by `event_type`).
    #[schema(value_type = Object)]
    pub payload: serde_json::Value,
}

#[derive(Serialize, ToSchema)]
pub struct HistoryResponse {
    pub game_id: String,
    pub version: i64,
    pub events: Vec<EventOut>,
}

/// Full event log for the current game, each with a human-readable description
/// and the version it produced. Use those versions with rewind/fork.
#[utoipa::path(get, path = "/game/history", tag = "Game", responses((status = 200, body = HistoryResponse)))]
async fn get_game_history(State(state): State<AppState>) -> AppResult<Json<HistoryResponse>> {
    let game = state.manager.get_state().await?;
    let events = state.manager.get_history().await?;
    let event_outs = events
        .into_iter()
        .map(|e| {
            // Full payload minus the "type" discriminator (matches Python).
            let mut payload = serde_json::to_value(&e.payload).unwrap_or(serde_json::Value::Null);
            if let Some(obj) = payload.as_object_mut() {
                obj.remove("type");
            }
            EventOut {
                version: e.sequence + 1,
                description: describe_event(&e.payload),
                event_type: e.payload.event_type().to_string(),
                timestamp: e.timestamp.to_rfc3339(),
                payload,
            }
        })
        .collect();
    Ok(Json(HistoryResponse {
        game_id: game.game_id.to_string(),
        version: game.version,
        events: event_outs,
    }))
}

#[derive(Deserialize, ToSchema)]
pub struct RewindRequest {
    /// Version to rewind to (1-based, inclusive).
    #[schema(example = 5)]
    pub to_version: i64,
}

/// Rewind the current game to a previous version (destructive — deletes later
/// events). Use `POST /game/fork` first to preserve the original timeline.
#[utoipa::path(
    post, path = "/game/rewind", tag = "Game",
    request_body = RewindRequest,
    responses(
        (status = 200, description = "Rewound state", body = GameStateResponse),
        (status = 400, description = "Version out of range (must be 1..=current)")
    )
)]
async fn rewind_game(
    State(state): State<AppState>,
    AppJson(req): AppJson<RewindRequest>,
) -> AppResult<Json<GameStateResponse>> {
    state.manager.rewind(req.to_version).await?;
    Ok(Json(build_game_state_response(&state).await?))
}

#[derive(Deserialize, ToSchema)]
pub struct ForkRequest {
    /// Version to fork from (1-based, inclusive).
    #[schema(example = 3)]
    pub from_version: i64,
}

#[derive(Serialize, ToSchema)]
pub struct ForkResponse {
    pub new_game_id: String,
    pub version: i64,
}

/// Fork the current game from a version into a new independent game (the
/// original is untouched). The forked game becomes active.
#[utoipa::path(
    post, path = "/game/fork", tag = "Game",
    request_body = ForkRequest,
    responses(
        (status = 200, description = "Forked game", body = ForkResponse),
        (status = 400, description = "Version out of range")
    )
)]
async fn fork_game(
    State(state): State<AppState>,
    AppJson(req): AppJson<ForkRequest>,
) -> AppResult<Json<ForkResponse>> {
    let new_state = state.manager.fork(req.from_version).await?;
    Ok(Json(ForkResponse {
        new_game_id: new_state.game_id.to_string(),
        version: new_state.version,
    }))
}

#[derive(Deserialize, ToSchema)]
pub struct LoadGameRequest {
    /// UUID of the game to load (from `GET /game/list`).
    pub game_id: String,
}

/// Load a previously saved game by id; it becomes the active game.
#[utoipa::path(
    post, path = "/game/load", tag = "Game",
    request_body = LoadGameRequest,
    responses(
        (status = 200, description = "Loaded state", body = GameStateResponse),
        (status = 400, description = "Invalid or unknown game id")
    )
)]
async fn load_game(
    State(state): State<AppState>,
    AppJson(req): AppJson<LoadGameRequest>,
) -> AppResult<Json<GameStateResponse>> {
    let game_id = Uuid::parse_str(&req.game_id)
        .map_err(|_| AppError::bad_request(format!("Invalid game ID: {}", req.game_id)))?;
    state.manager.load_game(game_id).await?;
    Ok(Json(build_game_state_response(&state).await?))
}

#[derive(Serialize, ToSchema)]
pub struct GameListResponse {
    pub game_ids: Vec<String>,
}

/// List all saved game ids. Use with `POST /game/load` to switch games.
#[utoipa::path(get, path = "/game/list", tag = "Game", responses((status = 200, body = GameListResponse)))]
async fn list_games(State(state): State<AppState>) -> AppResult<Json<GameListResponse>> {
    let ids = state.manager.list_games().await?;
    Ok(Json(GameListResponse {
        game_ids: ids.iter().map(|g| g.to_string()).collect(),
    }))
}

/// Stream game-state changes via Server-Sent Events.
///
/// Emits the full game state (the `GameStateResponse` minus `timer`) as a JSON
/// object in each `data:` frame — once immediately on connect, then again after
/// every mutation. Prefer this over polling `GET /game/state`.
#[utoipa::path(
    get, path = "/game/stream", tag = "Game",
    responses((
        status = 200,
        description = "SSE stream; each event's data is a game-state JSON object",
        content_type = "text/event-stream",
        body = GameStateBase
    ))
)]
async fn stream_game_state(
    State(state): State<AppState>,
) -> Sse<impl Stream<Item = Result<Event, Infallible>>> {
    let rx = state.manager.subscribe();
    let initial = state.manager.get_state().await.ok();
    let initial_event = initial.map(|s| Ok(Event::default().data(sse_payload(&s))));

    let updates = BroadcastStream::new(rx).filter_map(|res| match res {
        Ok(s) => Some(Ok(Event::default().data(sse_payload(&s)))),
        Err(_) => None,
    });

    let stream = tokio_stream::iter(initial_event).chain(updates);
    Sse::new(stream).keep_alive(KeepAlive::default())
}
